percona/controllers/pgc/pgc.go

```diff
@@ -0,0 +1,618 @@
+		deployment, err := c.Client.AppsV1().Deployments(pgCluster.Namespace).Get(ctx,
+			pgCluster.Name, metav1.GetOptions{})
+		if err != nil {
+			log.Errorf("could not find instance for pgcluster: %q", err.Error())
+			return
+		}
+		err = pmm.UpdatePMMSidecar(c.Client, pgCluster, deployment)
+		if err != nil {
+			log.Errorf("update pmm sidecar: %q", err.Error())
+		}
+		if _, err := c.Client.AppsV1().Deployments(deployment.Namespace).Update(ctx, deployment, metav1.UpdateOptions{}); err != nil {
+			log.Errorf("could not update deployment for pgcluster: %q", err.Error())
+		}
+	}
+
+	_, err = c.Client.CrunchydataV1().Pgclusters(keyNamespace).Update(ctx, pgCluster, metav1.UpdateOptions{})
+	if err != nil {
+		log.Errorf("update pgcluster resource: %s", err)
+		return
+	}
+}
+
+// onDelete is called when a pgcluster is deleted
+func (c *Controller) onDelete(obj interface{}) {
+	ctx := context.TODO()
+
+	// TODO: this object trick should be rework
+	clusterObj, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		log.Errorln("delete cluster: object is not DeletedFinalStateUnknown")
+		return
+	}
+	cluster, ok := clusterObj.Obj.(*crv1.PerconaPGCluster)
+	if !ok {
+		log.Errorln("delete cluster: object is not PerconaPGCluster")
+		return
+	}
+
+	err := c.Client.CrunchydataV1().Pgclusters(cluster.Namespace).Delete(ctx, cluster.Name, metav1.DeleteOptions{})
+	if err != nil {
+		log.Errorf("delete pgcluster resource: %s", err)
+	}
+}
+
+// AddPerconaPGClusterEventHandler adds the pgcluster event handler to the pgcluster informer
+func (c *Controller) AddPerconaPGClusterEventHandler() {
+	c.Informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onAdd,
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	})
+
+	log.Debugf("percona pgcluster Controller: added event handler to informer")
+}
+
+// WorkerCount returns the worker count for the controller
+func (c *Controller) WorkerCount() int {
+	return c.PerconaPGClusterWorkerCount
+}
+
+func createOrUpdateService(clientset kubeapi.Interface, cluster *crv1.PerconaPGCluster, svcType ServiceType) error {
+	ctx := context.TODO()
+	var service *corev1.Service
+	switch svcType {
+	case PGPrimaryServiceType:
+		svc, err := getPrimaryServiceObject(cluster)
+		if err != nil {
+			return errors.Wrap(err, "get primary service object")
+		}
+		service = svc
+	case PGBouncerServiceType:
+		svc, err := getPGBouncerServiceObject(cluster)
+		if err != nil {
+			return errors.Wrap(err, "get pgBouncer service object")
+		}
+		service = svc
+	}
+	oldSvc, err := clientset.CoreV1().Services(cluster.Namespace).Get(ctx, service.Name, metav1.GetOptions{})
+	if err != nil {
+		_, err = clientset.CoreV1().Services(cluster.Namespace).Create(ctx, service, metav1.CreateOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "create service %s", service.Name)
+		}
+		return nil
+	}
+	if reflect.DeepEqual(service.Spec, oldSvc.Spec) {
```

Comment:
in service spec you don't have the `Spec.ClusterIP` which will cause that this function will always return false